lib/spack/llnl/util/filesystem.py

```diff
@@ -974,6 +974,49 @@ def remove_linked_tree(path):
             shutil.rmtree(path, True)
 
 
+@contextmanager
+def safe_remove(*files_or_dirs):
+    """Context manager to remove the files passed as input, but restore
+    them in case any exception is raised in the context block.
+
+    Args:
+        *files_or_dirs: glob expressions for files or directories
+            to be removed
+
+    Returns:
+        Dictionary that maps deleted files to their temporary copy
+        within the context block.
+    """
+    # Find all the files or directories that match
+    glob_matches = [glob.glob(x) for x in files_or_dirs]
+
+    # Copy them in a temporary location
+    removed, dst_root = {}, tempfile.mkdtemp()
+
+    try:
+        for id, file_or_dir in enumerate(itertools.chain(*glob_matches)):
+            src = os.path.abspath(file_or_dir)
+            # The glob expression at the top ensures that the file/dir exists
+            # at the time we enter the loop. Double check here since it might
+            # happen that a previous iteration of the loop already removed it.
+            # This is the case, for instance, if we remove the directory
+            # "/foo/bar" before the file "/foo/bar/baz.yaml".
+            if not os.path.exists(src):
+                continue
+            # The monotonic ID is a simple way to make the filename
+            # or directory name unique in the temporary folder
+            basename = os.path.basename(file_or_dir) + '-{0}'.format(id)
+            dst = os.path.join(dst_root, basename)
+            shutil.move(src, dst)
+            removed[src] = dst
+        yield removed
+    except BaseException:
+        # Restore the files that were removed
+        for src, dst in removed.items():
```

Comment:
(minor) Could you use a name like `orig_path` instead of `src`, and `tmp_path` instead of `dst`? I think that would be slightly more clear but feel free to skip that if you disagree.

Reply:
Done in 8982ee4